### DISCUSSION TITLE
How to generate reports for HFSS Antenna Parameters?

### DISCUSSION THREAD

**Discussion by wsp6** [ansys/pyaedt#4949] [Q&A] [answered]
I'm attempting to use PyAEDT verstion 0.9 to generate a report for the radiation efficiency of an antenna.  It works great when done through the AEDT GUI, but I can't get a trace to show up when doing it through AEDT. I'm using this command:

        total_efficiency = hfss.post.create_report(
            expressions="dB(RadiationEfficiency)",
            report_category='Antenna Parameters',
            context="3D",
            domain='Sweep',
            primary_sweep_variable='Freq',
            plot_name="RadiationEfficiency",
            setup_sweep_name="Setupname : Sweepname",
        )

I get a rectangular plot labeled "RadiationEfficiency", but there is no trace.

I've tried multiple variations of the command based on this discussion: "https://github.com/ansys/pyaedt/discussions/4927" , but none of them seem to work for the antenna parameters category.

**Comment by Samuelopez-ansys** [comment, marked as answer]
Hi @wsp6 ,

This code should work:

```
import pyaedt

app = pyaedt.Hfss(version="2024.1", design="HFSSDesign1")

# Get report object
setups = app.existing_analysis_sweeps
infinite_spheres = app.field_setup_names
new_report = app.post.reports_by_category.antenna_parameters("dB(RadiationEfficiency)", 
                                                             infinite_sphere=infinite_spheres[0],
                                                             setup=setups[1])

# Here you can modify the new_report object. For instance:
# new_report.report_type = "Data Table"

# Create report
new_report.create()

# Get data
data = new_report.get_solution_data()
data.data_real()

app.release_desktop()
```

As you can see you can plot or get the data. "data" is an object that you can use to plot the trace in matplotlib or use for your own workflow.
![image](https://github.com/user-attachments/assets/0563afb5-f68e-47ee-bee3-75b00661e455)

Please close the issue if this helps you.

**Reply by wsp6** [reply]
Thanks for your help. I'm running into  a problem when trying to get the solution data.

This command : data = new_report.get_solution_data()
gives the error messages:
PyAEDT WARNING: Solution Data failed to load. Check solution, context or expression.
PyAEDT WARNING: No Data Available. Check inputs

The message manager in the GUI also has this error:
[error] Script macro error: No trace data for quantity: dB(RadiationEfficiency) (10:58:21 AM  Jul 31, 2024)

I'm running Ansys 2023 R2 with PyAEDT v0.9.9

I know the data is there because I can plot it through the GUI, here's the GUI produced plot.

![Antenna Params Plot 2](https://github.com/user-attachments/assets/731dfa30-8216-45aa-9d23-6e56ca725bcf)

**Reply by Samuelopez-ansys** [reply]
@wsp6 did the plot work?
 If not , probably you missed something in the creation of the new_report object. 
 Please share a project to reproduce the issue.

**Reply by wsp6** [reply]
The plot was created, I just can't populate it with a trace. Here's a screen shot of the results of the "new_report.create()" line

![Screenshot 2024-07-31 232459](https://github.com/user-attachments/assets/c91b573f-9774-4bd1-8d36-d76ebe7982fa)

How should I share my project? Should it be uploaded or emailed somewhere? Or do you want the *.aedt file added as an attachment to the forum post? 

Here's my code:

import pyaedt

hfss = pyaedt.Hfss(PROJ_PATH + PROJ_NAME, DESIGN_NAME)

for i in range(len(existing_setups)):
        hfss.delete_setup(existing_setups[i])

hfss.create_setup(
        name="PSOSetup",
        setup_type="HFSSDriven",
        Frequency="2.45GHZ",
        MaxDeltaS=0.01,
        MaximumPasses=20,
        BasisOrder=2,
        MinimumConvergedPasses=2,
    )

hfss.create_linear_count_sweep(
        setup="PSOSetup",
        name="PSOSweep",
        units="GHz",
        start_frequency=2.4,
        stop_frequency=2.5,
        num_of_freq_points=21,
        sweep_type="Interpolating",
        save_fields=True,
        save_rad_fields=True,
    )
    
hfss.parametrics.add_from_file(
        Path + "\Parametrics_Setup_" + ".csv",
        "Parametrics_Setup_",
        )
    
hfss.analyze_setup(
        name="Parametrics_Setup_",
        cores=4,
        gpus=1,
        tasks=4,
        blocking=True,
        use_auto_settings=False,
        allowed_distribution_types=["Variations", "Frequencies", "Mesh Assembly"],
    )

setups=hfss.existing_analysis_sweeps
infinite_spheres= hfss.field_setup_names

new_report = hfss.post.reports_by_category.antenna_parameters(
            "dB(RadiationEfficiency)",
            infinite_sphere=infinite_spheres[0],
            setup=setups[1]
            )
        
new_report.create()
test_data= new_report.get_solution_data()
test_data.data_real()

**Reply by Samuelopez-ansys** [reply]
@wsp6 you can drag and drop a ZIP file , please share the parametric csv too

**Reply by wsp6** [reply]
[Shared_project.zip](https://github.com/user-attachments/files/16459363/Shared_project.zip)

The attached zip has my AEDT file and the parametric file I'm using for testing.

**Comment by Vamsi-Yaddanapudi** [comment]
I am just wondering, if there is a need to "SaveFileds" even in the Parametrics setup.

**Comment by wsp6** [comment]
Edit: forgot to include the line of code where I set that option.  Here is how I'm setting the "SaveFields" option for the Parametrics setup. This line goes immediately after I import the parametrics setup from the CSV file.

hfss.set_oo_property_value(aedt_object=hfss.ooptimetrics, object_name="Parametrics_Setup_"), prop_name='SaveFields', value='True')


With this setting I can generate far field plots successfully. The following actions work
when I specify a particular variation from the parameters: create a polar
far field dB(gain) plot, export polar plot to jpg, create far field
dB(gain) table, and export far field gain table as csv.

Get_solution_data also partially works for dB(gain). I found an issue where
I can't get the information out of the exported object. The
solution_data.data_real() method of the object only outputs a one
dimensional vector instead of a two dimensional matrix. A separate issue
has already been created for this problem.

**Reply by Vamsi-Yaddanapudi** [reply]
Ok, I see you have another thread open. Do you still need some help for this thread?

**Comment by wsp6** [comment]
Yes. The antenna parameters values, specifically Radiation Efficiency, are better values for my application than the far field gain.

I'm currently exporting a far field gain data table to CSV, reading it back in, and then finding the max value. That other thread addresses a bug in the solution_data.data_real() function (which is why I have to write then immediately read a CSV).

That works, but the radiation efficiency would work better.

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by Samuelopez-ansys** [comment]
Hi @wsp6 ,

I have just launched your script, I think all the data is there. The problem is that by default PyAEDT creates the report with the Nominal variation, which in this case is not solved, you have solved only the parametric sweep.

```
setups = hfss.existing_analysis_sweeps
infinite_spheres = hfss.field_setup_names

new_report = hfss.post.reports_by_category.antenna_parameters(
    "dB(RadiationEfficiency)",
    infinite_sphere=infinite_spheres[0],
    setup=setups[1]
)

available_variations = hfss.available_variations.variations(output_as_dict=True)
dependent_variables = hfss.available_variations.nominal_w_values_dict

data = []
for variations in available_variations:
    # Delete dependent parameters
    independent_available_variations = {key: value for key, value in variations.items() if
                                        key in list(dependent_variables.keys())}
    new_report.variations = independent_available_variations

    new_report.create()

    test_data = new_report.get_solution_data()

    data.append(test_data.data_real())


hfss.release_desktop(False, False)
```
